addon/-private/wrap-field.js

```diff
@@ -0,0 +1,199 @@
+import Ember from 'ember';
+
+import {
+  isMandatorySetter,
+  isEmberDescriptor,
+  isDescriptorTrap
+} from './utils/computed';
+import { getPropertyDescriptor } from './utils/object';
+
+import { TypeError } from './errors';
+
+const notifyPropertyChange =
+  Ember.notifyPropertyChange || Ember.propertyDidChange;
+
+function guardBind(fn, ...args) {
+  if (typeof fn === 'function') {
+    return fn.bind(...args);
+  }
+}
+
+class ValidatedProperty {
+  constructor({ originalValue, klass, keyName, typeValidators }) {
```

Comment:
I was still getting used to native classes at the time of writing this. I think there are still benefits to passing an object of "named params" through, but I have a feeling it'll be problematic for TS if we ever switch. Either way, not a big deal at the moment, just adding context here